### ISSUE TITLE
`nntrainer::forwarding(input, label, training)` need attention

### ISSUE THREAD

**Issue by zhoonit** [nnstreamer/nntrainer#888]
1. `nntrainer::forwarding` is supposed **to be able to take less than batchsize** input, however, as the network graph is compiled beforehand, it does not have fixed value anymore (see validateInput(sharedConstTensors X))
   1. input, label reassigns the memory slot

This is one of major issue that deepq is failing

**Comment by kparichay** [comment]
> nntrainer::forwarding is supposed to be able to take less than batchsize input, however, as the network graph is compiled beforehand, it does not have fixed value anymore (see validateInput(sharedConstTensors X))

BatchSize is a value which takes crucial role in the calculation. Providing partially filled in the batches will result in wrong results. For example, consider batch normalization layer or loss layer (where result is directly dependent on the batch size). To handle this, `inference/train` sets the appropriate batch size before calling the `forward/backward`. `validateInput()` method is called after the batch size is update based on the batchsize provided as input and memory allocated updated as per the batch size.

> input, label reassigns the memory slot

For training, input, label memory must not be allocated as well, and must be received from the databuffer. This will be done as a part of #303.
This replacement must happen inference. The bug is that the manager must not allocate memory for them for inference. I will take a look at this.

**Comment by kparichay** [comment]
> nntrainer::forwarding is supposed to be able to take less than batchsize input

Inference gives the abstraction of being able to take variable size batch size input at the cost of memory allocations/deallocations with each run (batch size inside the layers must be updated as explained in the comment for the correct functioning). This can be optimized by hinting the optimizer to not keep memory allocated for bigger batch size and reuse it for smaller batch size but I am not sure how frequent is this scenario.
